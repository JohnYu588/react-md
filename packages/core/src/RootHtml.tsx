import {
  type HTMLAttributes,
  type HtmlHTMLAttributes,
  type ReactElement,
  type ReactNode,
} from "react";
import { type PropsWithRef } from "./types.js";

/**
 * @since 6.0.0
 */
export interface RootHtmlProps extends HtmlHTMLAttributes<HTMLHtmlElement> {
  /** @defaultValue `"ltr"` */
  dir?: string;

  /** @defaultValue `"en"` */
  lang?: string;

  /**
   * Any additional props to provide to the `<body>` element and also supports a
   * `ref` if that is required for some reason. Using a `ref` would make your
   * root layout a client component though.
   */
  bodyProps?: PropsWithRef<HTMLAttributes<HTMLBodyElement>, HTMLBodyElement>;

  /**
   * Convenience prop to replace `bodyProps={{ className: "custom-class-name" }}`.
   */
  bodyClassName?: string;

  /**
   * Any content to render before the `<body>` tag. This can be useful if you
   * need to render a custom `<head>` element.
   */
  beforeBodyChildren?: ReactNode;

  /**
   * Any content to render after the `<body>` tag. This can be useful to insert
   * `<script>` tags for analytics or other external libraries.
   */
  afterBodyChildren?: ReactNode;
}

/**
 * This is mostly for creating the root html for next.js applications that
 * defaults to setting the `dir="ltr"` and `lang="en"`. Additional props can be
 * passed to the root `<html>` and `<body>` elements and children can be
 * rendered before and after the `<body>` if needed.
 *
 * @example Simple Example
 * ```tsx
 * // src/app/layout.tsx
 * import { RootHtml } from "@react-md/core";
 * import { type ReactElement, type PropsWithChildren } from "react";
 * import { Roboto_Flex } from "next/font/google";
 *
 * const roboto = Roboto_Flex({
 *   subsets: ["latin"],
 *   variable: "--roboto",
 * });
 *
 * export default function RootLayout(props: PropsWithChildren): ReactElement {
 *   const { children } = props;
 *
 *   return <RootHtml className={roboto.variable}>{children}</RootHtml>;
 * }
 * ```
 *
 * @since 6.0.0
 */
export function RootHtml(props: RootHtmlProps): ReactElement {
  const {
    dir = "ltr",
    lang = "en",
    children,
    bodyProps,
    bodyClassName,
    beforeBodyChildren,
    afterBodyChildren,
    ...remaining
  } = props;

  return (
    <html dir={dir} lang={lang} {...remaining}>
      {beforeBodyChildren}
      <body {...bodyProps} className={bodyClassName || bodyProps?.className}>
        {children}
      </body>
      {afterBodyChildren}
    </html>
  );
}
