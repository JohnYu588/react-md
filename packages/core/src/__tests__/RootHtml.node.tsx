/** @jest-environment node */
/* eslint-disable testing-library/render-result-naming-convention */
import { describe, expect, it } from "@jest/globals";
import { renderToString } from "react-dom/server";

import { RootHtml } from "../RootHtml.js";

describe("RootHtml", () => {
  it("should render the html and body elements with the default props", () => {
    expect(renderToString(<RootHtml />)).toMatchSnapshot();
  });

  it("should allow a custom bodyClassName", () => {
    expect(
      renderToString(<RootHtml bodyClassName="body-class-name" />)
    ).toMatchSnapshot();
  });

  it("should allow the body className to be set through bodyProps when the bodyClassName prop is not provided", () => {
    const html1 = renderToString(
      <RootHtml bodyProps={{ className: "custom-body-class" }} />
    );
    const html2 = renderToString(
      <RootHtml
        bodyProps={{ className: "custom-body-name" }}
        bodyClassName="body-class"
      />
    );

    expect(html1).toContain("custom-body-class");
    expect(html1).toMatchSnapshot();

    expect(html2).not.toContain("custom-body-class");
    expect(html2).toMatchSnapshot();
  });

  it("should allow the lang to be overwritten", () => {
    expect(renderToString(<RootHtml lang="da" />)).toMatchSnapshot();
  });

  it("should allow the dir to be overwritten", () => {
    expect(renderToString(<RootHtml dir="rtl" />)).toMatchSnapshot();
  });

  it("should allow the dir and lang to be overwritten", () => {
    expect(renderToString(<RootHtml lang="ar" dir="rtl" />)).toMatchSnapshot();
  });
});
