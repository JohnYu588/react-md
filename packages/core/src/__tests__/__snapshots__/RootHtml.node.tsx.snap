// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RootHtml should allow a custom bodyClassName 1`] = `"<html dir="ltr" lang="en"><body class="body-class-name"></body></html>"`;

exports[`RootHtml should allow the body className to be set through bodyProps when the bodyClassName prop is not provided 1`] = `"<html dir="ltr" lang="en"><body class="custom-body-class"></body></html>"`;

exports[`RootHtml should allow the body className to be set through bodyProps when the bodyClassName prop is not provided 2`] = `"<html dir="ltr" lang="en"><body class="body-class"></body></html>"`;

exports[`RootHtml should allow the dir and lang to be overwritten 1`] = `"<html dir="rtl" lang="ar"><body></body></html>"`;

exports[`RootHtml should allow the dir to be overwritten 1`] = `"<html dir="rtl" lang="en"><body></body></html>"`;

exports[`RootHtml should allow the lang to be overwritten 1`] = `"<html dir="ltr" lang="da"><body></body></html>"`;

exports[`RootHtml should render the html and body elements with the default props 1`] = `"<html dir="ltr" lang="en"><body></body></html>"`;
